### ISSUE TITLE
On Large-Batch Training for Deep Learning: Generalization Gap and Sharp Minima

### ISSUE THREAD

**Issue by nohzen** [nohzen/paper_notes#2]
# 概要
バッチサイズを増やすと、汎化性能が低下することが知られているが、それがシャープなminimumに収束するためだと数値実験により確かめた。

# 論文情報
- 著者: N. S. Keskar(Northwestern University), _et al._ (intel含む)
- ICLR2017
- paper:
https://arxiv.org/abs/1609.04836
https://openreview.net/forum?id=H1oyRlYgg
- code: https://github.com/keskarnitish/large-batch-training

# 内容
## 背景
- Deep Learningの学習を並列化するには、バッチサイズを大きくする必要がある
- しかしながら、バッチサイズを大きくすると汎化性能が低下する[LeCun+(2012)]
- Lossのminimumは、フラットな方が汎化性能が高いことが知られている。
理由は、下の図から分かるようにフラットな方がズレに対してロバストだから/フラットなminimumの方が少ないパラメータで記述できるから
![flat_minimum](https://user-images.githubusercontent.com/29157627/68147325-29df3680-ff7d-11e9-8b9e-7b72b663f40a.png) 

## 実験
- optimizerにはADAMを用いた
- バッチサイズは
  - Large Batch: バッチサイズは学習データの10%
  - Small Batch: バッチサイズは256
- データセットはMNIST, TIMIT, CIFAR-10/100（画像分類）
- ネットワークはいくつかの種類のfully connected networkとCNN(AvexNet系)
- Lossのminimumのシャープさの定量化: Lossの2回微分を使いたいが、これは計算コストが高いので代替となる量(4)式を利用

### Large BatchとSmall BatchでのLossのシャープさの比較
Large Batch(LB)で学習した場合の収束先のパラメータと、Small Batch(SB)で学習した場合の収束先のパラメータを繋いだ時の、Loss(cross-entropy)やAccuracyの変化を調べた。
下図は、横軸がパラメータ(α=0がSmall Batch(SB)、α=1がLarge Batch(LB)に対応)、縦軸がLossとaccuracy。
F_1, F_2, C_1などはモデルの違い。
α=0のSBのminimumの方が、α=1のLBのminimumよりフラットであることが分かる。
<img src="https://user-images.githubusercontent.com/29157627/68324913-b7509100-010b-11ea-90f5-3f21124e40a9.png" width="600">

### バッチサイズを変化させた場合のシャープさや精度
下図のようにバッチサイズを大きくするほど、テストデータでの精度が下がり、minimumのシャープさが上がる。
<img src="https://user-images.githubusercontent.com/29157627/68325554-fe8b5180-010c-11ea-8e98-4677d7ac6594.png" width="600">

### Small batchで事前学習してからLarge batchで学習する
下図は横軸がSBで事前学習する際のエポック数、縦軸がテストデータでの精度とminimumのシャープさ。
SBで十分に事前学習すれば、LBの学習でも、minimumはフラットになり汎化性能も上がることが分かる。
LBの学習はシャープなminimumから抜けられないが、SBの学習はシャープなminimumから抜けフラットなminimumに行ける。LBの学習はフラットなminimum付近からスタートすれば、そのままフラットなminimumに留まる。
<img src="https://user-images.githubusercontent.com/29157627/68382533-38089f00-0197-11ea-87d2-8dc939baeff3.png" width="600">

### Lossとminimumのシャープさの変化
下図は、学習過程でのLossとシャープさの時間変化。
学習が進むとLossが下がり、図の右に移動する。すると、minimumのシャープさはSBの時にはフラットに、LBの時にはシャープになる。
<img src="https://user-images.githubusercontent.com/29157627/68383209-91bd9900-0198-11ea-8a14-6de95a0409b2.png" width="600">

### Appendix E. Large batchの学習の精度をあげる方法
- Data augmentation: large batchでData augmentationすると、small batch(data augmentationありでも)と精度が変わらなくなるが、minimumのシャープさはdata augmentationをしてもsmall batchほどにならない。
- conservative training: data augmentataionと同じで精度は良くなるが、minimumのシャープさはそこまで改善しない
- robust training (adversarial training): 汎化性能は改善しない 

# 感想/批判
- optimizerはADAMでなくSGDで検証して欲しかった（SGDの方がシンプルなのと、ADAMはSGDより汎化性能が悪いらしいので）
- Small batchで事前学習してから、Large batchで学習する方法だと、事前学習の部分がボトルネックとなって学習時間の短縮にはなりにくそう。
- Large batchの学習で、既にフラットなminimum付近にいたら、そのままフラットなminimumに収束できるなら、初期値が重要になりそう。初期値をフラットなminimumにする方法はない？（論文のdiscussionにも書かれている内容）

# 関連論文
バッチサイズを増やすと、汎化性能が低下する理由を理論的に研究した論文:
- [Jastrzebski+(2017)](https://arxiv.org/abs/1711.04623)
- [Simsekli+(2019)](https://arxiv.org/abs/1901.06053)
